File: MiscellaneousAppKit/MiscellaneousAppKit/Configuration/Model/ConfigurationSliderDescription.m
//
//  ConfigurationSliderDescription.m
//  MiscellaneousAppKit
//
//  Created by Jinwoo Kim on 2/1/25.
//

#import "ConfigurationSliderDescription.h"

@implementation ConfigurationSliderDescription

+ (ConfigurationSliderDescription *)descriptionWithSliderValue:(double)sliderValue minimumValue:(double)minimunValue maximumValue:(double)maximumValue {
    return [[[ConfigurationSliderDescription alloc] initWithSliderValue:sliderValue minimumValue:minimunValue maximumValue:maximumValue] autorelease];
}

+ (ConfigurationSliderDescription *)descriptionWithSliderValue:(double)sliderValue minimumValue:(double)minimunValue maximumValue:(double)maximumValue continuous:(BOOL)continuous {
    return [[[ConfigurationSliderDescription alloc] initWithSliderValue:sliderValue minimumValue:minimunValue maximumValue:maximumValue continuous:continuous] autorelease];
}

- (instancetype)initWithSliderValue:(double)sliderValue minimumValue:(double)minimunValue maximumValue:(double)maximumValue {
    return [self initWithSliderValue:sliderValue minimumValue:minimunValue maximumValue:maximumValue continuous:NO];
}

- (instancetype)initWithSliderValue:(double)sliderValue minimumValue:(double)minimunValue maximumValue:(double)maximumValue continuous:(BOOL)continuous {
    if (self = [super init]) {
        _sliderValue = sliderValue;
        _minimumValue = minimunValue;
        _maximumValue = maximumValue;
        _continuous = continuous;
    }
    
    return self;
}

- (id)copyWithZone:(struct _NSZone *)zone {
    return self;
}

- (ConfigurationSliderDescription *)configurationWithSliderValue:(double)sliderValue {
    return [ConfigurationSliderDescription descriptionWithSliderValue:sliderValue minimumValue:self.minimumValue maximumValue:self.maximumValue continuous:self.continuous];
}

@end

File: MiscellaneousAppKit/MiscellaneousAppKit/Configuration/Model/ConfigurationSliderDescription.swift
//
//  ConfigurationSliderDescription.swift
//  MiscellaneousAppKit
//
//  Created by Jinwoo Kim on 3/2/25.
//

import SwiftUI

public struct ConfigurationSliderDescription {
    private let impl: __ConfigurationSliderDescription
    
    private init(impl: __ConfigurationSliderDescription) {
        self.impl = impl
    }
    
    public init(value: Double, minValue: Double, maxValue: Double) {
        let impl = __ConfigurationSliderDescription(sliderValue: value, minimumValue: minValue, maximumValue: maxValue)
        self.init(impl: impl)
    }
    
    public init(value: Double, minValue: Double, maxValue: Double, continuous: Bool) {
        let impl = __ConfigurationSliderDescription(sliderValue: value, minimumValue: minValue, maximumValue: maxValue, continuous: continuous)
        self.init(impl: impl)
    }
}

extension ConfigurationSliderDescription: _ObjectiveCBridgeable {
    public func _bridgeToObjectiveC() -> __ConfigurationSliderDescription {
        impl
    }
    
    public static func _forceBridgeFromObjectiveC(_ source: __ConfigurationSliderDescription, result: inout ConfigurationSliderDescription?) {
        result = ConfigurationSliderDescription(impl: source)
    }
    
    public static func _conditionallyBridgeFromObjectiveC(_ source: __ConfigurationSliderDescription, result: inout ConfigurationSliderDescription?) -> Bool {
        result = ConfigurationSliderDescription(impl: source)
        return true
    }
    
    public static func _unconditionallyBridgeFromObjectiveC(_ source: __ConfigurationSliderDescription?) -> ConfigurationSliderDescription {
        ConfigurationSliderDescription(impl: source!)
    }
}

extension ConfigurationForm {
    public struct SliderItem: Item {
        public let _itemModel: ConfigurationItemModel?
        public let _didTriggerAction: (Any?) -> Void
        
        public init(
            identifier: String,
            title: String,
            value: Binding<Double>,
            minValue: Double,
            maxValue: Double
        ) {
            self.init(
                itemModel: ConfigurationItemModel
                    .sliderItem(
                        identifier: identifier,
                        label: title,
                        valueResolver: { itemModel in
                            ConfigurationSliderDescription(
                                value: value.wrappedValue,
                                minValue: minValue,
                                maxValue: maxValue
                            )
                        }
                    )
            ) { newValue in
                value.wrappedValue = newValue as! Double
            }
        }
        
        public init(
            identifier: String,
            title: String,
            value: Binding<Double>,
            minValue: Double,
            maxValue: Double,
            continuous: Bool
        ) {
            self.init(
                itemModel: ConfigurationItemModel
                    .sliderItem(
                        identifier: identifier,
                        label: title,
                        valueResolver: { itemModel in
                            ConfigurationSliderDescription(
                                value: value.wrappedValue,
                                minValue: minValue,
                                maxValue: maxValue,
                                continuous: continuous
                            )
                        }
                    )
            ) { newValue in
                value.wrappedValue = newValue as! Double
            }
        }
        
        private init(itemModel: ConfigurationItemModel, didTriggerAction: @escaping (Any?) -> Void) {
            _itemModel = itemModel
            _didTriggerAction = didTriggerAction
        }
    }
}
